File: test/leetcode/twosum/two_sum_test.ts
import chai from "https://cdn.skypack.dev/chai@4.3.4?dts";
import { TwoSum } from "../../../leetcode/twosum/two_sum.ts"

Deno.test("Two Sums Implementation", async (t)=>{
    // arrange
    const impl: TwoSum = new TwoSum();
    const expect = chai.expect;
    type testCase = {
        description: string;
        nums: Array<number>;
        target: number;
        expected: Array<number>;
    }
    const testCases: Array<testCase> = new Array<testCase>(
        {
            description: "Base Case A",
            nums: [2,7,11,15],
            target: 9,
            expected: [0,1],
        },
        {
            description: "Base Case B",
            nums: [3,2,4],
            target: 6,
            expected: [1,2],
        },
        {
            description: "Base Case C",
            nums: [3,3],
            target: 6,
            expected: [0,1],
        },
    )

    for (const tc of testCases) {
        await t.step(tc.description, ()=>{
            // act
            const actual = impl.Optimized(tc.nums, tc.target)
            // assert
            expect(actual).to.have.deep.members(tc.expected)
        })
    }
})

File: test/leetcode/twosum/two_sum_test.go
package twosum_test

import (
	"github.com/Ydot19/epi/leetcode/twosum"
	"sort"
	"testing"

	"github.com/stretchr/testify/suite"
)

type TwoSumSuite struct {
	suite.Suite
	Impl twosum.TwoSum
}

func (suite *TwoSumSuite) SetupTest() {
	suite.Impl = twosum.NewTwoSum()
}

func (suite *TwoSumSuite) Cases(a func(nums []int, target int) []int) {
	type testCases struct {
		description string
		nums        []int
		target      int
		expected    []int
	}
	for _, scenario := range []testCases{
		{
			description: "Base Case 1",
			nums:        []int{2, 7, 11, 15},
			target:      9,
			expected:    []int{0, 1},
		},
		{
			description: "Base Case 2",
			nums:        []int{3, 2, 4},
			target:      6,
			expected:    []int{1, 2},
		},
		{
			description: "Base Case 3",
			nums:        []int{3, 3},
			target:      6,
			expected:    []int{0, 1},
		},
	} {
		suite.T().Run(scenario.description, func(t *testing.T) {
			actual := a(scenario.nums, scenario.target)
			sort.Ints(actual)
			sort.Ints(scenario.expected)
			suite.Equal(scenario.expected, actual)
		})
	}

}

func (suite *TwoSumSuite) TestOptimized() {
	suite.Cases(suite.Impl.Optimized)
}

func TestTwoSum(t *testing.T) {
	suite.Run(t, new(TwoSumSuite))
}
